crates/goose/src/token_counter.rs

```diff
@@ -239,27 +215,32 @@ impl TokenCounter {
                     .unwrap_or_default()
                     .trim_end_matches('.');
                 let line = format!("{}:{}", name, description);
-                func_token_count += self.count_tokens(&line); // Add tokens for name and description
+                func_token_count += self.count_tokens(&line);
 
-                if let serde_json::Value::Object(properties) = &tool.input_schema["properties"] {
+                if let Some(serde_json::Value::Object(properties)) =
+                    tool.input_schema.get("properties")
+                {
                     if !properties.is_empty() {
-                        func_token_count += prop_init; // Add tokens for start of properties
+                        func_token_count += PROP_INIT;
                         for (key, value) in properties {
-                            func_token_count += prop_key; // Add tokens for each property
+                            func_token_count += PROP_KEY;
                             let p_name = key;
-                            let p_type = value["type"].as_str().unwrap_or("");
-                            let p_desc = value["description"]
-                                .as_str()
+                            let p_type = value.get("type").and_then(|v| v.as_str()).unwrap_or("");
+                            let p_desc = value
+                                .get("description")
+                                .and_then(|v| v.as_str())
                                 .unwrap_or("")
                                 .trim_end_matches('.');
                             let line = format!("{}:{}:{}", p_name, p_type, p_desc);
                             func_token_count += self.count_tokens(&line);
-                            if let Some(enum_values) = value["enum"].as_array() {
+                            if let Some(enum_values) = value.get("enum").and_then(|v| v.as_array())
+                            {
                                 func_token_count =
-                                    func_token_count.saturating_add_signed(enum_init); // Add tokens if property has enum list
+                                    func_token_count.saturating_add_signed(ENUM_INIT);
                                 for item in enum_values {
                                     if let Some(item_str) = item.as_str() {
-                                        func_token_count += enum_item;
+                                        func_token_count =
+                                            func_token_count.saturating_add_signed(ENUM_INIT);
```

Comment:
also here.

How is this different from the one in AsyncTokenCounter?

Reply:
this code is clearly duplicated